sql/src/main/java/org/apache/druid/sql/calcite/planner/querygen/DruidQueryGenerator.java

```diff
@@ -0,0 +1,296 @@
+ * under the License.
+ */
+
+package org.apache.druid.sql.calcite.planner.querygen;
+
+import com.google.common.collect.ImmutableList;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Aggregate;
+import org.apache.calcite.rel.core.Filter;
+import org.apache.calcite.rel.core.Project;
+import org.apache.calcite.rel.core.Sort;
+import org.apache.calcite.rel.core.Window;
+import org.apache.calcite.rex.RexBuilder;
+import org.apache.druid.error.DruidException;
+import org.apache.druid.query.QueryDataSource;
+import org.apache.druid.sql.calcite.planner.PlannerContext;
+import org.apache.druid.sql.calcite.planner.querygen.DruidQueryGenerator.PDQVertexFactory.PDQVertex;
+import org.apache.druid.sql.calcite.planner.querygen.InputDescProducer.InputDesc;
+import org.apache.druid.sql.calcite.rel.DruidQuery;
+import org.apache.druid.sql.calcite.rel.PartialDruidQuery;
+import org.apache.druid.sql.calcite.rel.PartialDruidQuery.Stage;
+
+import javax.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Converts a DAG of {@link org.apache.druid.sql.calcite.rel.logical.DruidLogicalNode} convention to a native
+ * {@link DruidQuery} for execution.
+ */
+public class DruidQueryGenerator
+{
+  private final RelNode relRoot;
+  private final PDQVertexFactory vertexFactory;
+
+  public DruidQueryGenerator(PlannerContext plannerContext, RelNode relRoot, RexBuilder rexBuilder)
+  {
+    this.relRoot = relRoot;
+    this.vertexFactory = new PDQVertexFactory(plannerContext, rexBuilder);
+  }
+
+  public DruidQuery buildQuery()
+  {
+    Vertex vertex = buildVertexFor(relRoot, true);
+    return vertex.buildQuery(true);
+  }
+
+  private Vertex buildVertexFor(RelNode node, boolean isRoot)
+  {
+    List<Vertex> newInputs = new ArrayList<>();
+    for (RelNode input : node.getInputs()) {
+      newInputs.add(buildVertexFor(input, false));
+    }
+    Vertex vertex = processNodeWithInputs(node, newInputs, isRoot);
+    return vertex;
+  }
+
+  private Vertex processNodeWithInputs(RelNode node, List<Vertex> newInputs, boolean isRoot)
+  {
+    if (node instanceof InputDescProducer) {
+      return vertexFactory.createVertex(PartialDruidQuery.create(node), newInputs);
+    }
+    if (newInputs.size() == 1) {
+      Vertex inputVertex = newInputs.get(0);
+      Vertex newVertex = inputVertex.mergeNode(node, isRoot);
+      if (newVertex != null) {
+        return newVertex;
+      }
+      inputVertex = vertexFactory.createVertex(
+          PartialDruidQuery.createOuterQuery(((PDQVertex) inputVertex).partialDruidQuery),
+          ImmutableList.of(inputVertex)
+      );
+      newVertex = inputVertex.mergeNode(node, false);
+      if (newVertex != null) {
+        return newVertex;
+      }
+    }
+    throw DruidException.defensive().build("Unable to process relNode[%s]", node);
+  }
+
+  /**
+   * Execution dag vertex - encapsulates a list of operators.
+   */
+  private interface Vertex
```

Comment:
Do we have any plans to extend this interface for other impls as well? Otherwise, could it be considered that we just have the private concrete class only?

Reply:
At some point I did had this as a stand-alone interface; but didn't wanted to retain something which only has one implementation - so moved it back inside the class.

I think keeping the private interface could help to move to multiple implementations later on as a simpler step.